### PR TITLE
Gives RCD a UI, and the ability to select airlock type and access

### PR DESCRIPTION
**Pull request by monster860**
Disclaimer: I did not give the normal RCD extra ammo capacity. I tested this with a combat RCD, which is identical to a normal one except it starts with 500 ammo.

:cl: monster860
rscadd: RCD gets a UI
rscadd: RCD-built airlocks can now receive different types and accesses.
/:cl:

### CODE REVIEW

**Comment by KasparoVy** [comment]
Pictures of the UI in action?

**Comment by monster860** [comment]
I already have pics.

**Comment by KasparoVy** [comment]
When I checked it was just the changelog, thanks for the pictures

**Comment by Twinmold93** [comment]
More advanced and useful RCD abilties? I see two sides to this: The usefulness in repair. And the major buff to antags who have an RCD, by effectively making doors that people can't follow them through with it.

**Comment by monster860** [comment]
@Twinmold93 The RCD can already build walls, and besides, doors are not exactly secure, any old fucker with a multitool, screwdriver, and crowbar can open a door in seconds.

**Comment by Earthdivine** [comment]
Does this get rid of the Sparks the RCD had? They were a good way to start a fire without being obvious like throwing a welding torch.

**Comment by Krausus** [comment]
This looks pretty handy, but it doesn't really make sense for basically anyone to be able to make limited-access doors, seeing as airlock electronics explicitly require engineering access to change their access requirements.

**Comment by Fox-McCloud** [comment]
>This looks pretty handy, but it doesn't really make sense for basically anyone to be able to make limited-access doors, seeing as airlock electronics explicitly require engineering access to change their access requirements.

This echos my biggest concern about it, as well.

**Comment by monster860** [comment]
I don't really get why the airlock electronics themselves require engineering access to edit doors. Any old fucker with tools can open them in seconds.

**Comment by TullyBurnalot** [comment]
The point here is that anyone with an RCD can now make Command-Level doors with a single click, @monster860 . Just because doors "can be hacked" does not mean giving this power to random crewmembers is a good idea.

**Comment by CrAzYPiLoT-SS13** [comment]
I don't really see the problem here. If a civvie gets ahold of an RCD and
creates a command-level airlock, what use of it does he have? He cannot
open it. Block himself off? Nope. Anyone can hack into an airlock with a
multitool, a crowbar and a screwdriver.

**Comment by FalseIncarnate** [comment]
Why not just make it so it requires you to "log in" or "authorize" an RCD to change door type (stays authorized until manually logged out) with an engi-access ID like the electronics do?

Swipe card on RCD, set type, print door, get disarmed and have a grey shirt print command doors everywhere because you didn't log out.

**Comment by Krausus** [comment]
> any old fucker with a multitool, screwdriver, and crowbar can open a door in seconds.

> Anyone can hack into an airlock with a multitool, a crowbar and a screwdriver.

Assuming that literally everyone is ready and willing to just hack their way through any airlock that gets in their way is nothing short of ludicrous.

> I don't really get why the airlock electronics themselves require engineering access to edit doors.

Because only engineers are supposed to be engineering doors.

> Why not just make it so it requires you to "log in" or "authorize" an RCD to change door type

For the same reason laser guns don't require you to swipe your security ID - the access controls should be on where tools are kept, not the tools themselves.

Which is probably the *real* problem with the RCD - it's a very sophisticated engineering device that some genius decided to make conveniently available from a (trivially) hacked autolathe. This PR would be fine if not for that. I wonder how much opposition there would be to removing RCDs from the autolathe...

**Comment by CrAzYPiLoT-SS13** [comment]
If only engineers are supposed to be making airlocks, remove the
functionality altogether. Heck, remove the RCD just for that reason.

**Comment by FlattestGuitar** [comment]
No, that's a good point. With this RCD you can remove any airlock and replace it with one of your choice, along with the correct access. This means that any non-engineer will be able to alter it, which is a balance change.

File: code/game/objects/items/weapons/RCD.dm [hunk 1]
```diff
@@ -26,232 +26,295 @@ RCD
...
+		/obj/machinery/door/airlock/glass_command = "Command (Glass)", /obj/machinery/door/airlock/glass_engineering = "Engineering (Glass)",
+		/obj/machinery/door/airlock/glass_security = "Security (Glass)", /obj/machinery/door/airlock/glass_medical = "Medical (Glass)",
+		/obj/machinery/door/airlock/mining = "Mining", /obj/machinery/door/airlock/atmos = "Atmospherics",
+		/obj/machinery/door/airlock/research = "Research", /obj/machinery/door/airlock/glass_research = "Research (Glass)",
+		/obj/machinery/door/airlock/glass_mining = "Mining (Glass)", /obj/machinery/door/airlock/glass_atmos = "Atmospherics (Glass)")
 
-	New()
+/obj/item/weapon/rcd/New()
+	desc = "A RCD. It currently holds [matter]/[max_matter] matter-units."
+	src.spark_system = new /datum/effect/system/spark_spread
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

File: code/game/objects/items/weapons/RCD.dm [hunk 2]
```diff
@@ -26,232 +26,295 @@ RCD
...
+	..()
+	if(istype(W, /obj/item/weapon/rcd_ammo))
+		var/obj/item/weapon/rcd_ammo/R = W
+		if((matter + R.ammoamt) > max_matter)
+			to_chat(user, "<span class='notice'>The RCD cant hold any more matter-units.</span>")
+			return
+		matter += R.ammoamt
+		user.drop_item()
+		qdel(W)
+		playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

File: code/game/objects/items/weapons/RCD.dm [hunk 3]
```diff
@@ -26,232 +26,295 @@ RCD
...
+		ui.set_auto_update(1)
 
-	proc/activate()
-		playsound(src.loc, 'sound/items/Deconstruct.ogg', 50, 1)
+/obj/item/weapon/rcd/Topic(href, href_list, nowindow, state)
+	if(..())
+		return 1
+	
+	if(prob(20))
+		src.spark_system.start()
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

File: code/game/objects/items/weapons/RCD.dm [hunk 4]
```diff
@@ -26,232 +26,295 @@ RCD
...
+			door_accesses -= href_access
+		else
+			door_accesses += href_access
+		door_accesses_list = list()
+		for(var/access in get_all_accesses())
+			door_accesses_list[++door_accesses_list.len] = list("name" = get_access_desc(access), "id" = access, "enabled" = (access in door_accesses))
+		. = 1
 
+/obj/item/weapon/rcd/proc/activate()
+	playsound(src.loc, 'sound/items/Deconstruct.ogg', 50, 1)
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

File: code/game/objects/items/weapons/RCD.dm [hunk 5]
```diff
@@ -26,232 +26,295 @@ RCD
...
+					activate()
+					var/turf/AT = A
+					AT.ChangeTurf(/turf/simulated/floor/plating)
+					return 1
+				return 0
+
+			if(istype(A, /turf/simulated/floor))
+				if(checkResource(3, user))
+					to_chat(user, "Building Wall ...")
+					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

File: code/game/objects/items/weapons/RCD.dm [hunk 6]
```diff
@@ -26,232 +26,295 @@ RCD
...
-							activate()
-							var/obj/machinery/door/airlock/T = new /obj/machinery/door/airlock( A )
-							T.autoclose = 1
-							return 1
-						return 0
+		if(2)
+			if(istype(A, /turf/simulated/floor))
+				if(checkResource(10, user))
+					to_chat(user, "Building Airlock...")
+					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

File: code/game/objects/items/weapons/RCD.dm [hunk 7]
```diff
@@ -26,232 +26,295 @@ RCD
...
-							var/turf/AT = A
-							AT.ChangeTurf(/turf/simulated/floor/plating)
-							return 1
+		if(3)
+			if(istype(A, /turf/simulated/wall))
+				if(istype(A, /turf/simulated/wall/r_wall) && !canRwall)
 					return 0
+				if(checkResource(5, user))
+					to_chat(user, "Deconstructing Wall...")
+					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

File: code/game/objects/items/weapons/RCD.dm [hunk 8]
```diff
@@ -26,232 +26,295 @@ RCD
...
-							if(!useResource(5, user)) return 0
-							activate()
-							var/turf/AT = A
-							AT.ChangeTurf(/turf/space)
-							return 1
-					return 0
+			if(istype(A, /turf/simulated/floor))
+				if(checkResource(5, user))
+					to_chat(user, "Deconstructing Floor...")
+					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

File: code/game/objects/items/weapons/RCD.dm [hunk 9]
```diff
@@ -26,232 +26,295 @@ RCD
...
-					A = locate(/obj/structure/grille) in A.loc
-				
-				if(istype(A, /obj/structure/grille))
-					if(!checkResource(2, user))
-						return 0
-					to_chat(user, "Deconstructing window...")
+			if(istype(A, /obj/machinery/door/airlock))
+				if(checkResource(20, user))
+					to_chat(user, "Deconstructing Airlock...")
 					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

File: code/game/objects/items/weapons/RCD.dm [hunk 10]
```diff
@@ -26,232 +26,295 @@ RCD
...
+				return	0
+				
+			if(istype(A, /obj/structure/window)) // You mean the grille of course, do you?
+				A = locate(/obj/structure/grille) in A.loc
+			
+			if(istype(A, /obj/structure/grille))
+				if(!checkResource(2, user))
+					return 0
+				to_chat(user, "Deconstructing window...")
+				playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

File: code/game/objects/items/weapons/RCD.dm [hunk 11]
```diff
@@ -26,232 +26,295 @@ RCD
...
+				return 1
+			return 0
+		if(4)
+			if(istype(A, /turf/simulated/floor))
+				if(locate(/obj/structure/grille) in contents)
+					return 0 // We already have window
+				if(!checkResource(2, user))
+					return 0
+				to_chat(user, "Constructing window...")
+				playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/RCD.dm]: src.

**Comment by TheDZD** [comment]
The access thing is still a problem. Otherwise, this looks okay.

**Comment by Fox-McCloud** [comment]
@monster860 going to make the access changes?

**Comment by monster860** [comment]
@Fox-McCloud Done

**Comment by FalseIncarnate** [comment]
:+1: with the new access change, makes RCDs much more useful for engineers.

**Pull request merged**